src/resolver.d.ts

```diff
@@ -0,0 +1,77 @@
+import * as Promise from 'bluebird'
+
+import Bundle from './bundle'
+import { FileInfo } from './fileInfo'
+
+// Make the external types available to implementers
+export { Bundle, FileInfo }
+/**
+ * Resolver: Base interface for any project type resolver.
+ */
+export interface Resolver {
+	/**
+	 * priority: The priority that this resolver should take, when trying to resolve
+	 * a resin-bundle, e.g. a resolver with a higher priority will attempt to make
+	 * sense of the given bundle before one with a lower priority
+	 */
+	priority: number
+
+	/**
+	 * name: The friendly name of the project that this resolver resolves
+	 */
+	name: string
+
+	/**
+	 * entry: Provide this resolver with a entry into a tar archive (the transport type
+	 * of a resin-bundle) and the resolver should save the contents if it is applicable
+	 * to this type of resolver. For example a Dockerfile.template resolver should save
+	 * the contents of a Dockerfile.template.
+	 *
+	 * @param file
+	 *	The contents and information about the file found.
+	 */
+	entry(file: FileInfo): void
+
+	/**
+	 * isSatisfied: Once all of the entries in the tar stream have been provided to
+	 * the resolvers, the isSatisfied function will be called in order of priority.
+	 * If a resolver returns true, that resolver will then be used to populate the
+	 * resin-bundle with the necessary files to pass to a docker build - generally
+	 * just a Dockerfile, but not limited to.
+	 *
+	 * @return
+	 *	True if this resolver has enough information to produce a docker-compatible
+	 *	build artifact (normally Dockerfile)
+	 */
+	isSatisfied(): boolean
+
+	/**
+	 * provideBundle: An implementor of this interface should take notice of the resin
+	 * bundle information passed in with this function. The resolver should not make
+	 * assumptions on when the bundle will be provided, with the only assurance being
+	 * that the bundle will be provided BEFORE isSatisfied(..) and resolve(..) are called.
+	 *
+	 * @param bundle
+	 *	A resin bundle
+	 */
+	provideBundle(bundle: Bundle): void
+
+	/**
+	 * resolve: Once a resolver has reported itself as being satisfied with the input,
+	 * resolve will be called, which will return a promise of a list of files to be added
+	 * to the bundle, which will allow Docker to build the bundle.
+	 *
+	 * @return
+	 *	A promise of a list of files which when added to the bundle allow docker
+	 *	to build the bundle
+	 */
+	resolve(): Promise<FileInfo[]>
+
+	/**
+	 * reset: This function will be called on all registered resolvers, after a resolve
+	 * has been complete (or failed). This means that a resolver can be used more than
+	 * once without having to instantiate a new instance.
+	 */
+	reset(): void
```

Comment:
This doesn't feel like something this module should require, let alone care about. We could just take resolvers, use them, and not require this method. We can leave this to downstream clients.

If somebody does want to write resolvers that need instantiating each time then there's no reason to stop them, and if they want to build resettable builders and reset them themselves, they can do that super easily by just waiting until the `resolveBundle` returns. I don't think this adds anything, it just constrains users.